File: 2096-step-by-step-directions-from-a-binary-tree-node-to-another/2096-step-by-step-directions-from-a-binary-tree-node-to-another.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public String getDirections(TreeNode root, int startValue, int destValue) {
        StringBuilder start=new StringBuilder();
        StringBuilder dest=new StringBuilder();
        boolean check=recursion(root,startValue,start);
        
        boolean check1=recursion(root,destValue,dest);
        
        int idx=0;
        while(idx<start.length() && idx<dest.length() && start.charAt(idx)==dest.charAt(idx))
        idx++;
        
        StringBuilder ans=new StringBuilder();
        for(int i=idx;i<start.length();i++)
        ans.append('U');
        
        ans.append(dest.substring(idx));
        return ans.toString();
    }
    private boolean recursion(TreeNode root,int value,StringBuilder sb)
    {
        if(root==null)
        return false;
        if(root.val==value)
        return true;
        sb.append('L');
        if(recursion(root.left,value,sb))
        return true;
        sb.deleteCharAt(sb.length()-1);
        sb.append('R');
        if(recursion(root.right,value,sb))
        return true;
        sb.deleteCharAt(sb.length()-1);
        
        return false;
    }
}

File: 2096-step-by-step-directions-from-a-binary-tree-node-to-another/2096-step-by-step-directions-from-a-binary-tree-node-to-another.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getDirections(self, root: Optional[TreeNode], startValue: int, destValue: int) -> str:
        def getPath(node, target, dir_to_node):
            if node.val == target:
                return dir_to_node
            if node.left:
                ch_Path = getPath(node.left, target, "L")
                if ch_Path is not None:
                    return dir_to_node + ch_Path
            if node.right:
                ch_Path = getPath(node.right, target, "R")
                if ch_Path is not None:
                    return dir_to_node + ch_Path
            return None

        sPath, dPath = getPath(root, startValue, ""), getPath(root, destValue, "")
        common = 0
        while common < min(len(sPath),len(dPath)) and sPath[common] == dPath[common]:
            common += 1
        return "U"*(len(sPath)-common) + dPath[common:]

